libbpf-tools/xfsslower.bpf.c

```diff
@@ -0,0 +1,163 @@
+// SPDX-License-Identifier: GPL-2.0
+// Copyright (c) 2020 Wenbo Zhang
+#include "vmlinux.h"
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_core_read.h>
+#include "xfsslower.h"
+
+#define BPF_F_INDEX_MASK		0xffffffffULL
+#define BPF_F_CURRENT_CPU		BPF_F_INDEX_MASK
+
+const volatile pid_t targ_tgid = 0;
+const volatile __u64 min_lat = 0;
+
+struct piddata {
+	u64 ts;
+	loff_t start;
+	loff_t end;
+	struct file *fp;
+};
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, 8192);
+	__type(key, u32);
+	__type(value, struct piddata);
+} start SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_PERF_EVENT_ARRAY);
+	__uint(key_size, sizeof(u32));
+	__uint(value_size, sizeof(u32));
+} events SEC(".maps");
+
+struct args {
+	__u64 args[5];
+	__u64 ret;
+};
+
+static __always_inline int
+probe_entry(struct file *fp, loff_t start, loff_t end)
+{
+	u64 id = bpf_get_current_pid_tgid();
+	struct piddata piddata = {};
+	u32 tgid = id >> 32;
+	u32 pid = id;
+
+	if (targ_tgid && targ_tgid != tgid)
+		return 0;
+
+	piddata.ts = bpf_ktime_get_ns();
+	piddata.fp = fp;
+	piddata.start = start;
+	piddata.end = end;
+	if (piddata.fp)
+		bpf_map_update_elem(&start, &pid, &piddata, 0);
+	return 0;
+}
+
+SEC("fentry/xfs_file_read_iter")
+int fentry__xfs_file_read_iter(struct args *ctx)
+{
+	struct kiocb *iocb = (void*)ctx->args[0];
+
+	return probe_entry(iocb->ki_filp, iocb->ki_pos, 0);
+}
+
+SEC("fentry/xfs_file_write_iter")
+int fentry__xfs_file_write_iter(struct args *ctx)
+{
+	struct kiocb *iocb = (void*)ctx->args[0];
+
+	return probe_entry(iocb->ki_filp, iocb->ki_pos, 0);
+}
+
+SEC("fentry/xfs_file_open")
+int fentry__xfs_file_open(struct args *ctx)
+{
+	struct file *fp = (void*)ctx->args[1];
+
+	return probe_entry(fp, 0, 0);
+}
+
+SEC("fentry/xfs_file_fsync")
+int fentry__xfs_file_fsync(struct args *ctx)
+{
+	struct file *fp = (void*)ctx->args[0];
+	loff_t start = (loff_t)ctx->args[1];
+	loff_t end = (loff_t)ctx->args[2];
+
+	return probe_entry(fp, start, end);
+}
+
+static __always_inline int probe_exit(struct args *ctx, char type)
+{
+	u64 id = bpf_get_current_pid_tgid();
+	u64 end_ns = bpf_ktime_get_ns();
+	struct piddata *piddatap;
+	struct event event = {};
+	struct qstr qs = {};
+	u32 tgid = id >> 32;
+	u32 pid = id;
+	u64 delta_us;
+
+	if (targ_tgid && targ_tgid != tgid)
+		return 0;
+
+	piddatap = bpf_map_lookup_elem(&start, &pid);
+	if (!piddatap)
+		return 0;    /* missed entry */
+
+	delta_us = (end_ns - piddatap->ts) / 1000;
+	bpf_map_delete_elem(&start, &pid);
+
+	if ((s64)delta_us < 0 || delta_us <= min_lat)
+		return 0;
+
+	qs = BPF_CORE_READ(piddatap, fp, f_path.dentry, d_name);
```

Comment:
given that you are using fentry/fexit programs, you should be able to do this equivalently as :

```
qs = pidatatap->fp->f_path.dentry->d_name;
```

how cool is that, eh? :)

Reply:
So cool! :) I found I missed 

> With tp_btf and fentry/fexit BPF program types, available since Linux 5.5, natural C syntax is possible as well.

from your blog.

Reply:
Actually, verifier might not allow direct memory access in this particular case, because you are storing `struct file *` in map, at which point verifier has no way to guarantee that this pointer won't be overwritten from userspace. Please try it, but you might need to stick to BPF_CORE_READ for this one, unfortunately.

Reply:
Thanks for the details, I'll try it later.